File: src/main/java/br/com/leonardoferreira/jirareport/domain/Issue.java
package br.com.leonardoferreira.jirareport.domain;

import br.com.leonardoferreira.jirareport.domain.embedded.Changelog;
import br.com.leonardoferreira.jirareport.domain.embedded.DueDateHistory;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.EqualsAndHashCode;
import lombok.NoArgsConstructor;
import lombok.ToString;
import org.hibernate.annotations.Type;

import javax.persistence.CascadeType;
import javax.persistence.Column;
import javax.persistence.Entity;
import javax.persistence.FetchType;
import javax.persistence.GeneratedValue;
import javax.persistence.GenerationType;
import javax.persistence.Id;
import javax.persistence.ManyToMany;
import javax.persistence.ManyToOne;
import javax.persistence.OneToMany;
import java.time.LocalDateTime;
import java.util.List;
import java.util.Map;
import java.util.Set;

/**
 * @author lferreira
 * @since 7/28/17 12:50 PM
 */
@Data
@Entity
@Builder
@NoArgsConstructor
@AllArgsConstructor
@ToString(of = "key")
@EqualsAndHashCode(callSuper = false)
public class Issue extends BaseEntity {
    private static final long serialVersionUID = -1084659211505084402L;

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    private String key;

    private String issueType;

    private String creator;

    private String system;

    private String epic;

    private String summary;

    private String estimated;

    private String project;

    private LocalDateTime startDate;

    private LocalDateTime endDate;

    private Long leadTime;

    private LocalDateTime created;

    private String priority;

    @Type(type = "jsonb")
    @Column(columnDefinition = "jsonb")
    private List<Changelog> changelog;

    @ManyToMany(mappedBy = "issues", cascade = CascadeType.DETACH)
    private List<IssuePeriod> issuePeriods;

    @OneToMany(mappedBy = "issue", fetch = FetchType.LAZY, cascade = CascadeType.ALL)
    private Set<LeadTime> leadTimes;

    @ManyToOne
    private Board board;

    private Long differenceFirstAndLastDueDate;

    @Type(type = "jsonb")
    @Column(columnDefinition = "jsonb")
    private List<DueDateHistory> dueDateHistory;

    private Long impedimentTime;

    @Type(type = "jsonb")
    @Column(columnDefinition = "jsonb")
    private Map<String, String> dynamicFields;

    private Long waitTime;

    private Long touchTime;

    private Double pctEfficiency;

}

File: src/main/java/br/com/jiratorio/domain/entity/Issue.kt
package br.com.jiratorio.domain.entity

import br.com.jiratorio.domain.entity.embedded.Changelog
import br.com.jiratorio.domain.entity.embedded.DueDateHistory
import lombok.AllArgsConstructor
import lombok.Builder
import lombok.Data
import lombok.EqualsAndHashCode
import lombok.NoArgsConstructor
import lombok.ToString
import org.hibernate.annotations.Type

import javax.persistence.CascadeType
import javax.persistence.Column
import javax.persistence.Entity
import javax.persistence.FetchType
import javax.persistence.GeneratedValue
import javax.persistence.GenerationType
import javax.persistence.Id
import javax.persistence.ManyToMany
import javax.persistence.ManyToOne
import javax.persistence.OneToMany
import java.time.LocalDateTime

@Entity
data class Issue(
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    var id: Long? = null,

    var key: String? = null,

    var issueType: String? = null,

    var creator: String? = null,

    var system: String? = null,

    var epic: String? = null,

    var summary: String? = null,

    var estimated: String? = null,

    var project: String? = null,

    var startDate: LocalDateTime? = null,

    var endDate: LocalDateTime? = null,

    @Column(nullable = false)
    val leadTime: Long? = null,

    var created: LocalDateTime? = null,

    var priority: String? = null,

    @Type(type = "jsonb")
    @Column(columnDefinition = "jsonb")
    var changelog: List<Changelog>? = null,

    @ManyToMany(mappedBy = "issues", cascade = [CascadeType.DETACH])
    var issuePeriods: List<IssuePeriod>? = null,

    @OneToMany(mappedBy = "issue", fetch = FetchType.LAZY, cascade = [CascadeType.ALL])
    var leadTimes: Set<LeadTime>? = null,

    @ManyToOne
    var board: Board? = null,

    var deviationOfEstimate: Long? = null,

    @Type(type = "jsonb")
    @Column(columnDefinition = "jsonb")
    var dueDateHistory: List<DueDateHistory>? = null,

    var impedimentTime: Long? = null,

    @Type(type = "jsonb")
    @Column(columnDefinition = "jsonb")
    var dynamicFields: Map<String, String>? = null,

    var waitTime: Long? = null,

    var touchTime: Long? = null,

    var pctEfficiency: Double? = null
) : BaseEntity() {
    companion object {
        private val serialVersionUID = -1084659211505084402L
    }

}
